feat(processPattern): add support for `{RegExp)` matches (`pattern.test`)

jkrems opened the pull request:
We want to match some of our existing asset logic which generates image file names from the source paths. Having a function that generates those paths seemed like the simplest solution.

This mirrors the way the `file-loader` works, allowing to reuse logic. https://github.com/webpack-contrib/file-loader#regexp

jkrems commented:
Actually, trying to make sure that templates are supported in the returned value.

michael-ciniawsky commented on README.md:
```diff
@@ -138,6 +138,9 @@ Or, in case of just a `from` with the default destination, you can also use a `{
       from: 'src/'
       to: 'dest/[name].[hash].[ext]',
       toType: 'template'
+      // Optional: provide a regExp to extract additional placeholders
```
Please make a new 'section'  for this option, or does it only work with `toType: 'template'` ?

michael-ciniawsky replied:
Also needs to be added to `options` table please

jkrems replied:
It will only work with the template `toType`. Will update the table.

jkrems commented:
Would it make sense to use `pattern.regExp` instead of `pattern.test`? That's consistent with how `file-loader` exposes this feature. Happy to switch to either. :)

jkrems commented:
* Added to options table
* Renamed to `options.test`
* Added dedicated section for `options.test` to the README containing the same docs that also appear in the `file-loader` README for `options.regExp`

evilebottnawi commented:
@jkrems rebase :+1:

jkrems commented:
Rebase'd.

michael-ciniawsky commented on README.md:
```diff
@@ -143,6 +144,27 @@ Or, in case of just a `from` with the default destination, you can also use a `{
 
+### `test`
+
+Defines a regExp to match some parts of the file path.
```
`regExp` => `{RegExp}`

jkrems replied:
Updated

michael-ciniawsky commented on README.md:
```diff
@@ -143,6 +144,27 @@ Or, in case of just a `from` with the default destination, you can also use a `{
+
+Defines a regExp to match some parts of the file path.
+These capture groups can be reused in the name property using `[N]` placeholder.
+Note that `[0]` will be replaced by the entire tested string,
```
```

> ℹ️ Note
```

`string` => `{String}`

jkrems replied:
Went with "entire relative path of the file" while I was touching it since that seemed easier to understand than what the "tested string" was.

michael-ciniawsky commented on README.md:
```diff
@@ -143,6 +144,27 @@ Or, in case of just a `from` with the default destination, you can also use a `{
+Defines a regExp to match some parts of the file path.
+These capture groups can be reused in the name property using `[N]` placeholder.
+Note that `[0]` will be replaced by the entire tested string,
+whereas `[1]` will contain the first capturing parenthesis of your regex
```
`regex` => `{RegExp}`

jkrems replied:
Done

jkrems commented:
Alright, the arguments to the interpolate call changed (relative -> absolute). Updated the test and docs to match that.

michael-ciniawsky commented:
Released in [`v4.5.0`](https://github.com/webpack-contrib/copy-webpack-plugin/releases/tag/v4.5.0) :tada:

jkrems commented:
Niiice! That was fast! Thanks a lot for the quick review & release. :)
